eras/shelley/impl/src/Cardano/Ledger/Shelley/CertState.hs

```diff
@@ -0,0 +1,148 @@
+    , shelleyCertDState = d
+    }
+
+shelleyCertDStateL :: Lens' (ShelleyCertState era) (DState era)
+shelleyCertDStateL = lens shelleyCertDState (\ds u -> ds {shelleyCertDState = u})
+{-# INLINE shelleyCertDStateL #-}
+
+shelleyCertPStateL :: Lens' (ShelleyCertState era) (PState era)
+shelleyCertPStateL = lens shelleyCertPState (\ds u -> ds {shelleyCertPState = u})
+{-# INLINE shelleyCertPStateL #-}
+
+shelleyCertVStateL :: Lens' (ShelleyCertState era) (VState era)
+shelleyCertVStateL = lens shelleyCertVState (\ds u -> ds {shelleyCertVState = u})
+{-# INLINE shelleyCertVStateL #-}
+
+toCertStatePairs :: KeyValue e a => ShelleyCertState era -> [a]
+toCertStatePairs ShelleyCertState {..} =
+  [ "dstate" .= shelleyCertDState
+  , "pstate" .= shelleyCertPState
+  ]
+
+shelleyObligationCertState :: ShelleyCertState era -> Obligations
+shelleyObligationCertState (ShelleyCertState VState {vsDReps} PState {psDeposits} DState {dsUnified}) =
+  let accum ans drepState = ans <> drepDeposit drepState
+   in Obligations
+        { oblStake = UM.fromCompact (UM.sumDepositUView (UM.RewDepUView dsUnified))
+        , oblPool = F.foldl' (<>) (Coin 0) psDeposits
+        , oblDRep = F.foldl' accum (Coin 0) vsDReps
+        , oblProposal = Coin 0
+        }
+
+shelleyCertsTotalDepositsTxBody ::
+  EraTxBody era => PParams era -> ShelleyCertState era -> TxBody era -> Coin
+shelleyCertsTotalDepositsTxBody pp ShelleyCertState {shelleyCertPState} =
+  getTotalDepositsTxBody pp (`Map.member` psStakePoolParams shelleyCertPState)
+
+shelleyCertsTotalRefundsTxBody ::
+  EraTxBody era => PParams era -> ShelleyCertState era -> TxBody era -> Coin
+shelleyCertsTotalRefundsTxBody pp ShelleyCertState {shelleyCertDState, shelleyCertVState} =
+  getTotalRefundsTxBody
+    pp
+    (lookupDepositDState shelleyCertDState)
+    (lookupDepositVState shelleyCertVState)
+
+instance EraCertState ShelleyEra where
+  type CertState ShelleyEra = ShelleyCertState ShelleyEra
+
+  mkCertState = mkShelleyCertState
+
+  upgradeCertState = error "It is not possible to translate Byron CertState with 'upgradeCertState'"
+
+  certDStateL = shelleyCertDStateL
+  {-# INLINE certDStateL #-}
+
+  certVStateL = shelleyCertVStateL
+  {-# INLINE certVStateL #-}
+
+  certPStateL = shelleyCertPStateL
+  {-# INLINE certPStateL #-}
+
+  obligationCertState = shelleyObligationCertState
+
+  certsTotalDepositsTxBody = shelleyCertsTotalDepositsTxBody
+
+  certsTotalRefundsTxBody = shelleyCertsTotalRefundsTxBody
+
+instance Era era => ToJSON (ShelleyCertState era) where
+  toJSON = object . toCertStatePairs
+  toEncoding = pairs . mconcat . toCertStatePairs
+
+instance Era era => EncCBOR (ShelleyCertState era) where
+  encCBOR ShelleyCertState {shelleyCertPState, shelleyCertDState, shelleyCertVState} =
+    encodeListLen 3
+      <> encCBOR shelleyCertVState
+      <> encCBOR shelleyCertPState
+      <> encCBOR shelleyCertDState
+
+instance Era era => DecShareCBOR (ShelleyCertState era) where
```

Comment:
Please take a look at the recent changes to all of the `DecShareCBOR` instances in this PR, since they have changed significantly since you started work on this work
See current instance for CertState: https://github.com/IntersectMBO/cardano-ledger/blob/bc10beb0038319354eefae31baf381193c5f4e32/libs/cardano-ledger-core/src/Cardano/Ledger/CertState.hs#L423-L439

Please, make sure you don't loose any of the functionality from this PR: https://github.com/IntersectMBO/cardano-ledger/pull/4826